cgroup.go

```diff
@@ -0,0 +1,50 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/docker/docker/pkg/parsers"
+)
+
+// set function in variable to overwrite for testing.
+var getCpusetGuest = func() string {
+	cpusetGuestByte, err := ioutil.ReadFile("/sys/fs/cgroup/cpuset/cpuset.cpus")
+	if err != nil {
+		fmt.Println("Error:", err)
+	}
+
+	return string(cpusetGuestByte)
+}
+
+// Return the best match for cpuset list in the guest.
+// The runtime caller may apply cpuset for specific CPUs in the host.
+// The CPUs may not exist on the guest as they are hotplugged based
+// on cpu and qouta.
+// This function return a working cpuset to apply on the guest.
+func getAvailableCpusetList(cpusetReq string) (string, error) {
+
+	cpusetGuest := getCpusetGuest()
+
+	cpusetListReq, err := parsers.ParseUintList(cpusetReq)
+	if err != nil {
+		return "", err
+	}
+
+	cpusetGuestList, err := parsers.ParseUintList(strings.TrimSpace(string(cpusetGuest)))
+	if err != nil {
+		return "", err
+	}
+
+	for k := range cpusetListReq {
+		if !cpusetGuestList[k] {
+			agentLog.Warnf("cpu %d is not in guest cpuset, using %s cpuset", k, cpusetGuest)
+			return cpusetGuest, nil
```

Comment:
`getAvailableCpusetList` returns all the cpus when a cpu in the list of cpuset is not available in the guest, for example if cpuset is 7, but the available cpus in the guest are 0-6, this function will return 0-6 even when only one cpu is requested. I wonder if this implementation is correct

Reply:
For me it is difficult to decide if correct or not. Because it is traying to make more compatible a broken use case in Kata that is a current limitation because we run the workload in VM with a different set of CPUs 

I prefer to not apply the cpuset at all instead of apply only where just a subset of them, why? 
- the workload is still limited to use limited cpu time based on cpu cgroup (cpu and qouta) so still be safe not consuming cpu time of other process. 

> for example if cpuset is 7, but the available cpus in the guest are 0-6, this function will return 0-6 even when only one cpu is requested

Lets translate to a docker example.
```
docker run -ti --runtime kata-runtime  --cpuset-cpus 7 --cpus 6  --net none --rm busybox
/ # cat /proc/self/status  | grep Cpus_allowed_list
Cpus_allowed_list:      0-5
```
1. On the host the VM will run only on this cpu, so the workload as well with  https://github.com/kata-containers/runtime/pull/1189
2. On the guest workload, the cpuset will allow to run the workload in all the vcpus (think cpuset on the host is still isolating our workload to jump to other process. 
3. From user perspective on the host this is still correct because the VM is only using only the CPU 7

Another  case:

```
docker run -tid --cpuset-cpu 1,5-9  --cpus 4 mycontainer
```
1. The  cpuset list is valid on the host and thanks to https://github.com/kata-containers/runtime/pull/1189 this is going to be limited on the host and the VM is allow to have its 4cpu time on any of the 1,5-9 host cpus (we isolate the workload)
2. In the guest we will have 4 cpus, and cpuset list of 1,5-9, where only cpu 1 is  available.

There was a few options. 
- Create a mapping for each cpu and vcpu, in this case is not even possible because host has at least 10 cpus and the guest hast 4. 
- Apply only the ones that are available, and ignore the rest, in this case the only cpu available is `1` so the workload would be able to run only on vcpu 1 but there rest of them wont be used by it, this mean waste of vcpus and the workload limited from the cpuset we partially apply to not get the cpu time of 4 cpus.
- If the request has cpus that are not part of it, just ignore it. The cpuset is not applied on the guest and workload will be jumping around the guest vcpus but the host cpuset will prevent the VM going to not allowed host cpus, and the workload will  be still limited to use 4 cpus because its cpu cgroup ( period and quota). 

So I went for the last option where it if is possible the cpuset is applied. But if not there is still isolation at host level. The solution is not perfect but solve a very basic docker use case and works with cpu cgroup and cpuset host cgroup  to work in a more compatible way.
```
docker run -tid --cpuset-cpu 2-4  --cpus 2 mycontainer
```
And should be the same for kubernetes when it isolate workloads to use a set of cpus. I still need to check this use case.

Reply:
The cpuset itself needs to be translated before passing into the guest, imo.  As it is provided, it is only valid for physical CPUs.  Once cpuset is done on the host (has this landed already?), then this PR becomes relevant.  IMO, after the runtime pins on the host, the cpuset information needs to be translated to map from physical CPUs to virtual CPUs in the guest.  We shouldn't just pass the original request in.

Reply:
@egernst agree. I work on follow up PRs to do this, this is only allowing to run Kata on K8s static policy and docker.